daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala

```diff
@@ -1387,14 +1381,15 @@ trait ElementBaseGrammarMixin
       }
       case LengthKind.Implicit if isSimpleType && primType == PrimType.String =>
         new SpecifiedLengthImplicitCharacters(this, body, this.maxLength.longValue)
-
-      case LengthKind.Implicit if isSimpleType && primType == PrimType.HexBinary =>
-        new SpecifiedLengthImplicit(this, body, this.maxLength.longValue * bitsMultiplier)
       case LengthKind.Implicit
-          if isSimpleType && impliedRepresentation == Representation.Binary =>
+          if isSimpleType &&
+            impliedRepresentation == Representation.Binary &&
+            primType != PrimType.HexBinary =>
         new SpecifiedLengthImplicit(this, body, implicitBinaryLengthInBits)
-      case LengthKind.Implicit if isComplexType =>
+      case LengthKind.Implicit =>
         body // for complex types, implicit means "roll up from the bottom"
+      // for simple types, the primitives have custom parsers that handle implicit length logic
+      // and don't use the limit provided by the SpecifiedLengthImplicit parser
```

Comment:
Can you move this comment above `body`? Single line comments can be on the same line, but multi-line comments should start before the line.